src/back/auth/src/service/service.cpp

```diff
@@ -156,6 +162,94 @@ model::UserInfo Service::GetUserInfo(const std::string& sessionId)
 	return _oidc.GetUserInfo(session._accessToken);
 }
 
+model::SessionRefresh Service::RefreshSession(
+	const std::string& sessionId,
+	const std::string& userAgent)
+{
+	// todo: need to refactor everything here, it's just a raw prototype
+
+	// проверка на наличие такой же сессии, но неактивной
+	SameInactiveSessionSecurityCheck(sessionId);
+	
+	// Получаем сессию из базы
+	auto session = _session.Table().GetById(sessionId, true);
+
+	// проверка на несовпадение userAgent
+	DifferentDeviceSecurityCheck(userAgent, session._device);
+
+	// Обновляем OIDC токены если требуется
+	if (_tokenizer.IsExpired(session._accessToken))
+	{
+		// TODO: dest lock!
+		updateTokens(session);
+	}
+	
+	// Обновляем токен Сессии
+
+	std::string token = updateSession(session, userAgent);
+
+	model::SessionRefresh refresh = {
+		._token = token
+	};
+
+	return refresh;
+}
+
+void Service::DifferentDeviceSecurityCheck(
+	const std::string& currentUserAgent,
+	const std::string& oldUserAgent
+){
+	if (strcmp(currentUserAgent.c_str(), oldUserAgent.c_str()) != 0){
+		LOG_WARNING() << "DIFFERENT USER AGENT SECURITY RISK!";
+		throw errors::ForceBlock{};
+	}
+}
+
+void Service::SameInactiveSessionSecurityCheck(const std::string& sessionId)
+{
+	bool isSecurityRisk = true;
+	model::Session check = {};
+	try{
+		check = _session.Table().GetById(sessionId, false); 
+	} catch(const std::exception& e) {
+		if (strcmp(e.what(), "resource not found") == 0){
+			isSecurityRisk = false;
+		} else {
+			throw;
+		}
+	}
```

Comment:
Такое можно было бы сделать вот так:
```suggestion
	} catch(const NotFound& e) {
		isSecurityRisk = false;
	}
```

Reply:
Поправил.